openslides_backend/action/actions/topic/json_upload.py

```diff
@@ -0,0 +1,138 @@
+
+
+@register_action("topic.json_upload")
+class TopicJsonUpload(DuplicateCheckMixin, Action):
+    """
+    Action to allow to upload a json. It is used as first step of an import.
+    """
+
+    model = Topic()
+    schema = DefaultSchema(Topic()).get_default_schema(
+        additional_required_fields={
+            "data": {
+                "type": "array",
+                "items": {
+                    "type": "object",
+                    "properties": {
+                        **model.get_properties("title", "text"),
+                        **{
+                            prop: agenda_creation_properties[prop]
+                            for prop in (
+                                "agenda_comment",
+                                "agenda_type",
+                                "agenda_duration",
+                            )
+                        },
+                    },
+                    "required": ["title"],
+                    "additionalProperties": False,
+                },
+                "minItems": 1,
+                "uniqueItems": False,
+            },
+            "meeting_id": required_id_schema,
+        }
+    )
+    permission = Permissions.AgendaItem.CAN_MANAGE
+
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        data = instance.pop("data")
+
+        # enrich data with meeting_id
+        for entry in data:
+            entry["meeting_id"] = instance["meeting_id"]
+
+        # generate headers
+        self.headers = [
+            {"property": "title", "type": "string"},
+            {"property": "text", "type": "string"},
+            {"property": "agenda_comment", "type": "string"},
+            {"property": "agenda_type", "type": "string"},
+            {"proptery": "agenda_duration", "type": "number"},
+        ]
+
+        # validate
+        self.init_duplicate_set(instance["meeting_id"])
+        self.rows = [self.validate_entry(entry) for entry in data]
+
+        # generate statistics
+        itemCount, itemNew, itemError = 0, 0, 0
+        for entry in self.rows:
+            itemCount += 1
+            if entry["status"] == "new":
+                itemNew += 1
+            if entry["status"] == "error":
+                itemError += 1
+        self.statistics = {
+            "itemCount": itemCount,
+            "Created": itemNew,
+            "Error": itemError,
+        }
+
+        # store rows in the action_worker
+        self.new_store_id = self.datastore.reserve_id(collection="action_worker")
+        fqid = fqid_from_collection_and_id("action_worker", self.new_store_id)
+        self.datastore.write_action_worker(
+            WriteRequest(
+                events=[
+                    Event(
+                        type=EventType.Create,
+                        fqid=fqid,
+                        fields={
+                            "id": self.new_store_id,
+                            "result": self.rows,
```

Comment:
It is not necessary, because there is a type-specific action, i.e. topic.import, for this

Reply:
Yes, and `topic.import` only gets an action worker id. If in the action worker it is saved to which import it belongs, the validation is easier that no invalid id was provided.

Reply:
I think it is a good idea, to include an import: "topic" in the action_worker, so that the import action can look, that has correct data. I include it.